---
title: "Structural variation in purple sea urchins"
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
editor_options: 
  markdown: 
    wrap: 72
---

# Using local PCA on the urchins

# 1. Input data processing

The pipeline will start from filtered bcf files, one for each of the 21
chromosomes.

Below is the steps that I followed to get from the raw sequence files to
these bcf files.

## Mapped reads to the reference

### The reference genome

<https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000002235.5/>

N50: 37.3 Mb

### The mapping algorithm

Input: List of read files (R1 and R2)

```{bash, eval = FALSE}
while read line ; do
        F1=$(cut -d ' ' -f1 <<< $line)
        F2=$(cut -d ' ' -f2 <<< $line)
        echo "$F1 -- $F2"
        FILE=$(mktemp)
        cat header.txt >> $FILE
        echo "spack load samtools@1.10" >> $FILE
        echo "spack load bwa@0.7.17" >> $FILE
        ref="/users/c/p/cpetak/WGS/reference_genome/GCF_000002235.5_Spur_5.0_genomic.fna"
        out_name=$(cut -d '.' -f1 <<< $F1)
        echo "bwa mem -t 1 -M $ref /users/c/p/cpetak/WGS/all_fastqs/$F1 /users/c/p/cpetak/WGS/all_fastqs/$F2 | samtools view -S -b > /users/c/p/cpetak/WGS/BWA_out/$out_name.bam" >> $FILE
        sbatch $FILE
        sleep 0.5
        rm $FILE
done < $1
```

The Burrows-Wheeler Alignment Tool (BWA) MEM algorithm was used for
mapping the raw reads to the S. purpuratus reference genome (Spur ver.
5.0, scaffold N50 ∼37 Mbp). The average coverage for each individual was
6.42±0.78, with an average mapping rate of 81.6±0.01.

## Called variants for each chromosome across all individuals

Input:

-   21 chromosome names
-   list_of_files.txt, 140 lines, line 1:
    `/users/c/p/cpetak/WGS/BWA_out/BOD_18170X61_200925_A00421_0244_AHKML5DSXY_S81_L002_R1_001.rmdup.bam`

```{bash, eval = FALSE}
while read line ; do
	echo "$line"
	FILE=$(mktemp)
  cat header.txt >> $FILE
  ref="/users/c/p/cpetak/WGS/reference_genome/GCF_000002235.5_Spur_5.0_genomic.fna"
  echo "echo "${line}" " >> $FILE
  echo "bcftools mpileup -r $line -f $ref --bam-list list_of_files.txt | bcftools call -mv -Ob -o multi_bam_${line}.bcf" >> $FILE
  sbatch $FILE
  sleep 0.5
  rm $FILE
done < $1
```

## Filtering the bcf files

```{bash, eval = FALSE}
#!/bin/sh

mychr="NW_022145603.1"

# use bcftools to filter your chromosome. The output of this line will be a vcf file that we can look at

bcftools view -e 'QUAL <= 40 || DP < 560 || MQB < -3 || RPB < -3 || RPB > 3 || AN < 238' ~/mydata/str_data/multi_bam_${mychr}.bcf > ~/mydata/str_data/${mychr}_filtered.vcf

echo "Filtered bcf" # Some printing to keep track of progress

# Convert the filtered vcf into the bcf file type which is the type the R package will be expecting

bcftools view -Ob ~/mydata/str_data/${mychr}_filtered.vcf > ~/mydata/str_data/${mychr}_filtered.bcf

echo "Converted to bcf" # Some printing to keep track of progress

# Index the filtered bcf file. This will make the file more searchable by the algorythm reading it.

bcftools index ~/mydata/str_data/${mychr}_filtered.bcf
echo "Indexed bcf"
echo "Done!"
```

Output: bcf files in the
`/users/c/p/cpetak/EG2023/structural_variation/backup/filtered_bcf_index`
directory.

# 2 Local PCA results

INPUT: name of the chromosome

OUTPUT:

-   Local PCA MDS plot with chromosome plot + PCAs for the 3 corners for
    bp [*testing*] and SNP sizes [500,1000,5000,10_000]
-   mds_coords.csv (2D MDS plot coordinates), mychromosome.regions.csv
    (genomic position of windows), pca_corners

Local PCA is an R package, here is the GitHub page:
<https://github.com/petrelharp/local_pca>.

Installation:

```         
install.packages("data.table")
devtools::install_github("petrelharp/local_pca/lostruct")
library(lostruct)
```

## 2.1 Run lostruct

On the github page, they provide an Rscript to show how to use the
package. I copied this:
[run_lostruct.R](https://github.com/Cpetak/local_pca_pipe/blob/main/run_lostruct.R)

No changes, expect output folder has chromosome name instead of random
id.

```         
#conda activate grn
#spack load bcftools@1.10.2

echo "Chromosome number: $1"

input_dir="/users/c/p/cpetak/EG2023/structural_variation/backup/filtered_bcf_index/${1}"

echo $input_dir

for snp in 500 1000 5000 10000
do
    echo "$snp"
    FILE=$(mktemp)
    cat header.txt >> $FILE
    echo "Rscript ~/WGS/local_pca_pipe/run_lostruct.R -i ${input_dir} -t snp -s ${snp} -I ~/WGS/local_pca_pipe/sample_info.tsv -c ${1}" >> $FILE
    sbatch $FILE
    #cat $FILE
    sleep 0.1
    rm $FILE
done

#for bp in $(seq 500 1000 5000 10000)
#do
    #echo "$bp"
    #FILE=$(mktemp)
    #cat header.txt >> $FILE
    #echo "Rscript ~/WGS/local_pca_pipe/run_lostruct.R -i ${input_dir} -t bp -s ${bp} -I ~/WGS/local_pca_pipe/sample_info.tsv -c ${1}" >> $FILE
    #sbatch $FILE
    #cat $FILE
    #sleep 0.1
    #rm $FILE
#done
```

To run: `bash 2_1_local_pca.sh NW_022145594.1`

# 2.2 Visualise and gather data

They also have an example of plotting the data, I copied this as well.
[summarize_run.Rmd](https://github.com/Cpetak/local_pca_pipe/blob/main/summarize_run.Rmd)

Changes:

-   do.pdfs \<- TRUE in line 14,

-   commented out warning in line 78, breaks code otherwise

-   added saving of corner pcas

-   added getting the percent corner (i.e. alpha) as a variable read
    from a file

Once the 2.1 script finishes, you should have a folder called
lostruct_results and a folder in it, named
type_snp_size_10000_chromosome_NW_022145594.1.

```         
#conda activate grn
#spack load bcftools@1.10.2

echo "Chromosome number: $1"
echo "Type: $2"
echo "Size: $3"
echo "Percent corner: $4"

input_dir="/users/c/p/cpetak/EG2023/structural_variation/backup/filtered_bcf_index/${1}"

echo $input_dir

cd ~/WGS/local_pca_pipe/lostruct_results/type_${2}_size_${3}_chromosome_${1}
echo $4 > percent_file.txt #the R script below will be looking for this file! easier than figuring out how to pass in as an argument

cd ~/WGS/local_pca_pipe

FILE=$(mktemp)
cat header.txt >> $FILE
echo "Rscript -e 'templater::render_template(\"~/WGS/local_pca_pipe/summarize_run.Rmd\",output=\"~/WGS/local_pca_pipe/lostruct_results/type_${2}_size_${3}_chromosome_${1}/run_summary.html\",change.rootdir=TRUE)'" >> $FILE
sbatch $FILE
#cat $FILE
sleep 0.1
rm $FILE
```

To run: `bash 2_2_local_pca_vis.sh NW_022145594.1 snp 10000 0.01`

It is chromosome, type, size, alpha (percent outliers considered). To
run for every snp length independently, a one-line:

`for snp in 500 1000 5000 10000; do bash 2_2_local_pca_vis.sh NW_022145594.1 snp $snp 0.05; done`
Then can rerun specific ones to adjust alpha.

# 3. Finding regions of interest

INPUT: MDS output files for a specific chromosome, for a specific type
of window size (snp vs bp)

Python file, finding_outlier_windows.py

To run: `python finding_outlier_windows.py NW_022145594.1 snp`

You can add an optional argument `--thr 0.2` to adjust the MDS outlier
threshold.

OUTPUT: Outlier region genomic coordinates for a specific chromosome for
each window size tested

# 4. Grouping individuals by genotype for specific region

INPUT: Coordinates of region of interest, bcf file of chromosome

Now that we have regions of interest (to recap: local pca -\> mds values
-\> outlier region coordinates) we should first look at the PCA of the
individuals based on that region. Yes, technically, this is what local
PCA is doing already, but it is cleaner and safer to start fresh from
the bcf files and do our own, independent PCA. The result should be the
same or similar! it might not be exactly the same because the PCAs that
are plotted as part of the local PCA pipe line only include outlier
windows, whereas here we are going to take all SNPs in a specific region
for the PCA analysis, and this might include windows that were not
outliers in the local PCA analysis.

Step 1: Make a vcf file that only includes the region of interest for
our chromosome.

run: `bash subset_bcf.sh NW_022145594.1 12670717 16440127` where it is
chromosome, start, end. Looks for a vcf file in the common data folder
for that chromosome (EG2023), output is also a vcf in the curr directory

Step 2: Convert vcf to gds

run:
`Rscript vcf2gds.R NW_022145594.1_12670717_16440127.vcf NW_022145594.1_12670717_16440127`
input filename, output file name (it will add .gds)

Step 3: Do PCA from the gds file

run: `Rscript do_pca.R NW_022145594.1_12670717_16440127.gds` does PCA
and writes PC1 and PC2 to csv files.

Step 4: Assign individuals to genotype groups, make plots

run:
`python genotype_by_PCA.py NW_022145594.1_12670717_16440127_dim1.csv NW_022145594.1_12670717_16440127_dim2.csv`,
output is 2 csv files with individual IDs (1-140) for each homozygote
group and PCA, pie and map plots, and chi-square results to txt file

**PUT TOGETHER**

run: `bash 4_local_pca.sh NW_022145594.1 12670717 16440127`

OUT: AA, Aa or aa for each individual and PCA plot, with map

# 4.2 Fst between homozygotes

INPUT: Individual IDs for the two homozygote groups, bcf file of the
chromosome, determined in 3.

(Need `vcf_col_ids` which is a list of column names in the vcf file
corresponding to the individuals. Can be generated by
`bcftools query -l NW_022145594.1_filtered.vcf > vcf_col_ids`)

Step 1:

`bcftools@1.10.2`

run: `bash extract_inds_from_vcf.sh NW_022145594.1_12670717_16440127`,
output is NW_022145594.1_12670717_16440127_vcf_list_homop and
NW_022145594.1_12670717_16440127_vcf_list_homoq which are lists of vcf
column names to keep for each group

Step 2:

`spack load vcftools@0.1.14`

`vcftools --gzvcf ~/EG2023/structural_variation/backup/filtered_vcf/NW_022145594.1_filtered.vcf --weir-fst-pop NW_022145594.1_12670717_16440127_vcf_list_homoq --weir-fst-pop NW_022145594.1_12670717_16440127_vcf_list_homop --out NW_022145594.1_12670717_16440127_fst`

**PUT TOGETHER**

run: `bash 4_2_local_pca.sh NW_022145594.1 12670717 16440127`

OUTPUT: Fst value for each position in the chromosome

# 5. GO enrichment

# 6. Whole chromosome analysis

## A. Linkage disequilibrium

### Triangle LD plot, aka window-based approach

Step 1: Chromosome vcf file to gds

Run:
`Rscript vcf2gds.R ~/EG2023/structural_variation/backup/filtered_vcf/NW_022145594.1_filtered.vcf NW_022145594.1_filtered`,
same script from before, makes gds file - pre-computed for all
chromosomes, in folder TODO

Step 2: Calculate LD with a certain window size

run: `sbatch makegrid_launcher.sh NW_022145594.1_filtered.gds 100000`,
file name, window size, submits 200 jobs and calls makegrid.R. Output is
200 files, each containing a subset of the windows.

Step 3: Put together the 200 files

run: `Rscript gather_results.R NW_022145594.1 100000` , output is
combined_chrom_windowsize.Rdata

run: `rm -rf ~/WGS/local_pca_pipe/makegrid_${1}_${2}`

Step 4: Make triangle plot

run: `Rscript plot_LD.R NW_022145594.1 100000`, output is a jpg

**PUT TOGETHER (given having .gds)**

`bash 6_a_local_pca.sh NW_022145594.1 100000`

### Triangle and U-plot for specific region

For zooming in a specific region, repeat above but with a subsetted vcf

> ***NOTE:*** Above code can be used that same but file names only
> include chromosome and LD window size so make sure to move output
> figures to avoid overwriting them!

For U-plot:

``` r
library(ggplot2) 
load("~/WGS/local_pca_pipe/combined_30k_job.Rdata")

#Get rid of NAs 
o <- na.omit(o)

#All window average r2 
ggplot(data=o, aes(x=abs(mid1-mid2), y=meanR2), alpha=.5, size=.5) + geom_point()
```

## B. Nucleotide diversity

# 7. Making a report

Rscript -e "rmarkdown::render('testing.Rmd', params=list(chr =
'NW_022145594.1', start = '12670717', stop = '16440127'))"
