---
title: "Structural variation in purple sea urchins"
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---

# Using local PCA on the urchins


# 1. Input data processing
The pipeline will start from filtered bcf files, one for each of the 21 chromosomes.

Below is the steps that I followed to get from the raw sequence files to these bcf files.

## Mapped reads to the reference

### The reference genome

<https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000002235.5/>

N50: 37.3 Mb

### The mapping algorythm

Input: List of read files (R1 and R2)

```{bash, eval = FALSE}
while read line ; do
        F1=$(cut -d ' ' -f1 <<< $line)
        F2=$(cut -d ' ' -f2 <<< $line)
        echo "$F1 -- $F2"
        FILE=$(mktemp)
        cat header.txt >> $FILE
        echo "spack load samtools@1.10" >> $FILE
        echo "spack load bwa@0.7.17" >> $FILE
        ref="/users/c/p/cpetak/WGS/reference_genome/GCF_000002235.5_Spur_5.0_genomic.fna"
        out_name=$(cut -d '.' -f1 <<< $F1)
        echo "bwa mem -t 1 -M $ref /users/c/p/cpetak/WGS/all_fastqs/$F1 /users/c/p/cpetak/WGS/all_fastqs/$F2 | samtools view -S -b > /users/c/p/cpetak/WGS/BWA_out/$out_name.bam" >> $FILE
        sbatch $FILE
        sleep 0.5
        rm $FILE
done < $1
```

The Burrows-Wheeler Alignment Tool (BWA) MEM algorithm was used for mapping the raw reads to the S. purpuratus reference genome (Spur ver. 5.0, scaffold N50 ∼37 Mbp). The average coverage for each individual was 6.42±0.78, with an average mapping rate of 81.6±0.01.

## Called variants for each chromosome across all individuals

Input:

-   21 chromosome names
-   list_of_files.txt, 140 lines, line 1: `/users/c/p/cpetak/WGS/BWA_out/BOD_18170X61_200925_A00421_0244_AHKML5DSXY_S81_L002_R1_001.rmdup.bam`

```{bash, eval = FALSE}
while read line ; do
	echo "$line"
	FILE=$(mktemp)
  cat header.txt >> $FILE
  ref="/users/c/p/cpetak/WGS/reference_genome/GCF_000002235.5_Spur_5.0_genomic.fna"
  echo "echo "${line}" " >> $FILE
  echo "bcftools mpileup -r $line -f $ref --bam-list list_of_files.txt | bcftools call -mv -Ob -o multi_bam_${line}.bcf" >> $FILE
  sbatch $FILE
  sleep 0.5
  rm $FILE
done < $1
```

## Filtering the bcf files

```{bash, eval = FALSE}
#!/bin/sh

mychr="NW_022145603.1"

# use bcftools to filter your chromosome. The output of this line will be a vcf file that we can look at

bcftools view -e 'QUAL <= 40 || DP < 560 || MQB < -3 || RPB < -3 || RPB > 3 || AN < 238' ~/mydata/str_data/multi_bam_${mychr}.bcf > ~/mydata/str_data/${mychr}_filtered.vcf

echo "Filtered bcf" # Some printing to keep track of progress

# Convert the filtered vcf into the bcf file type which is the type the R package will be expecting

bcftools view -Ob ~/mydata/str_data/${mychr}_filtered.vcf > ~/mydata/str_data/${mychr}_filtered.bcf

echo "Converted to bcf" # Some printing to keep track of progress

# Index the filtered bcf file. This will make the file more searchable by the algorythm reading it.

bcftools index ~/mydata/str_data/${mychr}_filtered.bcf
echo "Indexed bcf"
echo "Done!"
```

Output: bcf files in the `/users/c/p/cpetak/EG2023/structural_variation/backup/filtered_bcf_index ` directory.

# 2 Local PCA results
INPUT: name of the chromosome

OUTPUT: 
  - Local PCA MDS plot with chromosome plot + PCAs for the 3 corners for bp [] and snp sizes [500,1000,5000,10_000]
  - mds_coords.csv, mychromosome.regions.csv

Local PCA is an R package, here is the GitHub page: <https://github.com/petrelharp/local_pca>. 

Installation:
```
install.packages("data.table")
devtools::install_github("petrelharp/local_pca/lostruct")
library(lostruct)
```

## 2.1 Run lostruct

On the github page, they provide an Rscript to show how to use the package. I copied this: [run_lostruct.R](https://github.com/Cpetak/local_pca_pipe/blob/main/run_lostruct.R)

No changes, expect output folder has chromosome name instead of random id.

```
#conda activate grn
#spack load bcftools@1.10.2

echo "Chromosome number: $1"

input_dir="/users/c/p/cpetak/EG2023/structural_variation/backup/filtered_bcf_index/${1}"

echo $input_dir

for snp in 500 1000 5000 10000
do
	echo "$snp"
	FILE=$(mktemp)
  	cat header.txt >> $FILE
  	echo "Rscript ~/WGS/local_pca_pipe/run_lostruct.R -i ${input_dir} -t snp -s ${snp} -I ~/WGS/local_pca_pipe/sample_info.tsv -c ${1}" >> $FILE
    sbatch $FILE
  	#cat $FILE
    sleep 0.1
  	rm $FILE
done

#for bp in $(seq 500 1000 5000 10000)
#do
	#echo "$bp"
	#FILE=$(mktemp)
  	#cat header.txt >> $FILE
  	#echo "Rscript ~/WGS/local_pca_pipe/run_lostruct.R -i ${input_dir} -t bp -s ${bp} -I ~/WGS/local_pca_pipe/sample_info.tsv -c ${1}" >> $FILE
    #sbatch $FILE
  	#cat $FILE
    #sleep 0.1
  	#rm $FILE
#done
```
To run: `bash 2_1_local_pca.sh NW_022145594.1`

# 2.2 Visualise and gather data

They also have an example of plotting the data, I copied this as well.
[summarize_run.Rmd](https://github.com/Cpetak/local_pca_pipe/blob/main/summarize_run.Rmd)

Changes: 
- do.pdfs <- TRUE in line 14, 
- commented out warning in line 78, breaks code otherwise
- added saving of corner pcas
- added getting the percent corner (i.e. alpha) as a variable read from a file

Once the 2.1 script finishes, you should have a folder called lostruct_results and a folder in it, named type_snp_size_10000_chromosome_NW_022145594.1.


`Rscript -e 'templater::render_template("~/myscripts/summarize_run.Rmd",output="~/mydata/str_data/lostruct_results/type_snp_size_1000_weights_none_jobid_166584/run_summary.html",change.rootdir=TRUE)'`



# 3. Finding regions of interest
INPUT: MDS output files for a specific chromosome
OUTPUT: Outlier region genomic coordinates

# 4. Further analysis of region
INPUT: Genomic coordinates of interest
OUTPUT: PCA for that region, Fst, LD, nuc diversity, GO
