{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17000.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>589.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>188.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>189.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>504.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2112.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9900.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0\n",
       "0  17000.000\n",
       "1      1.000\n",
       "2    589.000\n",
       "3    188.000\n",
       "4    189.000\n",
       "5      0.000\n",
       "6    504.000\n",
       "7   2112.000\n",
       "8      0.375\n",
       "9   9900.000"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr=\"610\"\n",
    "df=pd.read_csv(f\"~/Desktop/snpEff/{chr}.csv\", header=None)\n",
    "df_inv=pd.read_csv(f\"~/Desktop/snpEff/{chr}_inv.csv\", header=None)\n",
    "df_inv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>409522.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>271.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51469.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20044.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20093.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>160.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>44624.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>87.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.4503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>156915.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0\n",
       "0  409522.0000\n",
       "1     271.0000\n",
       "2   51469.0000\n",
       "3   20044.0000\n",
       "4   20093.0000\n",
       "5     160.0000\n",
       "6   44624.0000\n",
       "7      87.0000\n",
       "8       0.4503\n",
       "9  156915.0000"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[  270 50880 19856 19904   160 44120]\n",
      "Variants rate:  87.0  vs  2112.0\n",
      "Missense / Silent ratio:  0.4503  vs  0.375\n",
      "high_impact\n",
      "Expected : 7.808331330674889\n",
      "Got: 1\n",
      "0.9995954066809903\n",
      "0.003567270209289151\n",
      "0.006482615858436354\n",
      "----------------------\n",
      "low_impact\n",
      "Expected : 1471.436659647179\n",
      "Got: 589\n",
      "1.0\n",
      "6.29527845456179e-180\n",
      "1.2649065664919188e-179\n",
      "----------------------\n",
      "moderate_impact\n",
      "Expected : 574.2304700069651\n",
      "Got: 188\n",
      "1.0\n",
      "1.5475330935589315e-83\n",
      "2.6176834974417633e-83\n",
      "----------------------\n",
      "missense\n",
      "Expected : 575.6186177990852\n",
      "Got: 189\n",
      "1.0\n",
      "1.8090404132391717e-83\n",
      "3.754891695852303e-83\n",
      "----------------------\n",
      "nonsense\n",
      "Expected : 4.627159307066601\n",
      "Got: 0\n",
      "1.0\n",
      "0.009767763526699396\n",
      "0.017847326503186815\n",
      "----------------------\n",
      "silent\n",
      "Expected : 1275.9391789236151\n",
      "Got: 504\n",
      "1.0\n",
      "1.6588104619381599e-155\n",
      "3.0094284716066287e-155\n",
      "----------------------\n"
     ]
    }
   ],
   "source": [
    "total_snps = int(df.iloc[0][0])\n",
    "inversion_snps = int(df_inv.iloc[0][0])\n",
    "total_outside = total_snps - inversion_snps\n",
    "\n",
    "total_intergenic = int(df.iloc[-1][0])\n",
    "inversion_intergenic = int(df_inv.iloc[-1][0])\n",
    "outside_intergenic = total_intergenic - inversion_intergenic\n",
    "\n",
    "total_outside = total_outside - outside_intergenic\n",
    "inversion_snps = inversion_snps - inversion_intergenic\n",
    "\n",
    "values_total = np.array(df.iloc[1:7][0].tolist()).astype(int)\n",
    "values_inverted = np.array(df_inv.iloc[1:7][0].tolist()).astype(int)\n",
    "values_outside = values_total - values_inverted\n",
    "print(values_outside)\n",
    "\n",
    "value_labels = [\"high_impact\", \"low_impact\", \"moderate_impact\", \"missense\", \"nonsense\", \"silent\"]\n",
    "\n",
    "#took out modifier: \"Usually non-coding variants or variants affecting non-coding genes, where predictions are difficult or there is no evidence of impact.\"\n",
    "\n",
    "print(\"Variants rate: \", df.iloc[-3][0], \" vs \", df_inv.iloc[-3][0])\n",
    "print(\"Missense / Silent ratio: \", df.iloc[-2][0], \" vs \", df_inv.iloc[-2][0])\n",
    "\n",
    "expected_values = []\n",
    "for i in range(len(values_total)):\n",
    "    print(value_labels[i])\n",
    "    prob = (values_outside[i])/total_outside #probability of high impact in the background\n",
    "    n = inversion_snps #number of coins I flipped, each SNP in inversion could be high impact\n",
    "    k = values_inverted[i] #successfully finding high impact\n",
    "\n",
    "    if k >= n:\n",
    "        print(\"less SNP - intergenic than impact of interest number of SNPs!\")\n",
    "        #NOTE num SNPs per impact level don't add up to total number of SNPs, > than that\n",
    "        k = n\n",
    "\n",
    "    print(f\"Expected : {n*prob}\")\n",
    "    print(f\"Got: {k}\")\n",
    "\n",
    "    expected_values.append(n*prob)\n",
    "\n",
    "    result=stats.binomtest(k, n, prob, alternative='greater') # less means what are the chances that k is less than n * p\n",
    "    print(result.pvalue)\n",
    "    result=stats.binomtest(k, n, prob, alternative='less') # less means what are the chances that k is less than n * p\n",
    "    print(result.pvalue)\n",
    "    result=stats.binomtest(k, n, prob, alternative='two-sided') # less means what are the chances that k is less than n * p\n",
    "    print(result.pvalue)\n",
    "\n",
    "    print(\"----------------------\")\n",
    "\n",
    "# High impact: more than expected! 4.65 vs 15, p < 0.001\n",
    "# Low impact: less than expected! 500.34 vs 369, p < 0.001\n",
    "# Moderate impact: more than expected! 268.28 vs 545, p < 0.001\n",
    "\n",
    "# Missense: more than expected! 269.11 vs 550, p < 0.001     #different amino acids\n",
    "# Nonsense: more than expected! 2.5 vs 6, p = 0.042          #premature stop\n",
    "# Silent: less than expected! 436.62 vs 327, p < 0.001       #no effect on the amino acid\n",
    "\n",
    "#two-sided p-vals were also significant\n",
    "\n",
    "# In the paper: \"Ignoring intergenic regions to avoid confounding effects from gene density\"...\n",
    "# As in, could have more missense because there are more SNPs in genes due to higher gene density.\n",
    "# However, in our case, inversion has more SNPs that are intergenic actually, 51.7% vs 26.1 % in the whole chromosome. \n",
    "# Exonic in inversion is 6.6%, for the whole 8%. so there are less candidate SNPs in inversion yet there are more that are missense. \n",
    "# included in above calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wgs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
